designs/operator-monitoring-plugin.md

```diff
@@ -0,0 +1,225 @@
+---
+title: controller-observation-plugin
+authors:
+  - "@Kavinjsir"
+reviewers:
+  - "@varsha"
+  - "@camila"
+approvers:
+creation-date: 2022-06-06
+last-updated: 2022-06-06
+status: provisional
+see-also:
+  - ""
+replaces:
+  - ""
+superseded-by:
+  - ""
+---
+
+# a-grafana-dashboard-plugin-to-observe-controller-metrics
+
+## Release Signoff Checklist
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Open Questions [optional]
+
+- 1. What does the ideal approach for the plugin to create the grafana manifests?
+- 2. Should we scaffold a dir with the config to install the grafana? See: https://grafana.com/docs/grafana/latest/installation/kubernetes/
+- 3. Where should the default directory be scaffold?
+  - grafana/
+  - grafana/dashboards
+  - grafana/config
+  - OR inside of the config dir another grafana dir with the k8s manifests to be applied on the cluster when the plugin is used
+- 4. How is the plugin backwards compatible for users who have prometheus scaffolded?
+  - Is it compatible only with specific versions of prometheus? Or do can we have an existing kb project and add this plugin to scaffold grafana manifests smoothly. Do we need any other changes to an old project structure.
+
+## Summary
+
+This EP will provide a way for the operator author to visualize the runtime metrics in order to observe the operator status.
+
+We currently have controller-runtime metrics exposed by default for all KB operators at an endpoint through Prometheus.
+
+In order to visualize those same metrics, every operator author has to individually put in the effort to create a dashboard.
+
+This plugin intends to reduce the effort of operator authors to create dashboards and instead provides with the manifests that will allow them to integrate their controller metrics with grafana easily.
+
+On top of the basic dashboard provided by us having default metrics, operator authors can also add custom metrics and visualize them in the format they are comfortable in.
+
+The functionality will be implemented by a new plugin that can generate the 'manifests' to be loaded on Grafana.
+
+In the context of this EP, the Grafana 'manifests' are mainly:
+
+- JSON files to be directly loaded on Grafana dashboards
+- TBD
+
+## Motivation
+
+The [controller-runtime](https://github.com/kubernetes-sigs/controller-runtime) provides rich metrics to measure controllers' status from various perspectives.
+
+Currently, users of Kubebuilder need to design and create their own Grafana dashboards to visualize the metrics.
+It would be great if we can provide a unified and comprehensive approach so that users may:
+
+- easily visualize controller/operator metrics
+- avoid duplicate creation of Grafana dashboards file
+- improve the observability towards the opeartor, keep it safe and robust
```

Comment:
WDYT about 
```suggestion

Kubebuilder and SDK users are looking for how to monitor their solutions using Grafana and the Prometheus metrics. Therefore, they are looking for helpers to speed up their solutions and to know how they can do it for their own customized metrics. For further information check some issues raised:

- [Grafana dashboard for visualizing controller-runtime metrics #2183](https://github.com/kubernetes-sigs/kubebuilder/issues/2183)
- [Custom resource metrics #2286](https://github.com/kubernetes-sigs/kubebuilder/issues/2286)

Therefore for SDK users and for those that are looking to distribute their solution via [OLM](https://olm.operatorframework.io/) in the [OperatorHub.io](https://operatorhub.io/) or via vendors like Openshift this new helper plugin will help them out achieve further [Operator Capability levels](https://sdk.operatorframework.io/docs/overview/operator-capabilities/) since for Operator projects are able to achieve the ultimate levels Deep Insight and auto-pilot the must provide mechanisms to allow cluster admins Monitoring their solutions. 
```